src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoNotNullable.cs

```diff
@@ -22,6 +22,10 @@ public override void Read(JsonTokenType tokenType, ref ReadStack state, ref Utf8
             {
                 // Forward the setter to the value-based JsonPropertyInfo.
                 JsonPropertyInfo propertyInfo = ElementClassInfo.GetPolicyProperty();
+                if (state.Current.IsProcessingDictionary && state.Current.KeyName == null)
```

Comment:
Since this code path is perf-sensitive can you please put this logic in `JsonSerializer.HandlePropertyName` instead (which will only get executed for dictionary property names).

Reply:
Per offline discussion, that is not possible. At least move this to ReadEnumerable where there is an existing check. Also you may need to check if IsProcessingImmutableDictionary and add a test for that (cc @layomia)